man/man1/openpgp2ssh.1
.\"  -*- nroff -*-
.Dd $Mdocdate: January 18, 2013 $
.Dt OPENPGP2SSH 1
.Os
.Sh NAME
openpgp2ssh
.Nd translate OpenPGP keys to SSH keys
.Sh SYNOPSIS
.Nm openpgp2ssh < mykey.gpg 
.Pp
.Nm gpg \-\-export $KEYID | openpgp2ssh $KEYID
.Pp
.Nm gpg \-\-export $KEYID | openpgp2pem $KEYID
.Pp
.Nm gpg \-\-export $KEYID | openpgp2spki $KEYID
.Pp
.Nm gpg \-\-export\-secret\-key $KEYID | openpgp2ssh $KEYID
.Sh DESCRIPTION
.Nm
takes an OpenPGP-formatted primary key and associated
subkeys on standard input, and spits out the requested equivalent
SSH-style (or PEM-encoded) key on standard output.
.Pp
If the data on standard input contains no subkeys, you can invoke
.Nm
without arguments.  If the data on standard input contains multiple
keys (e.g. a primary key and associated subkeys), you must specify a
specific OpenPGP key identifier as the first argument to indicate
which key to export.  The key ID is normally the 40 hex digit OpenPGP
fingerprint of the key or subkey desired, but
.Nm
will accept as few as the last 8 digits of the fingerprint as a key
ID.
.Pp
If the input contains an OpenPGP RSA public key, it will be converted
to the OpenSSH-style single-line keystring, prefixed with the key type
(`ssh\-rsa').  This format is suitable (with minor alterations) for
insertion into known_hosts files and authorized_keys files.  If
invoked as `openpgp2pem', a PEM-encoded public key will be emitted
instead.
.Pp
If invoked as `openpgp2spki', a PEM-encoded subjectPublicKeyInfo (as
defined in the X.509 standard) will be emitted instead.
.Pp
If the input contains an OpenPGP RSA secret key, it will be converted
to the equivalent PEM-encoded private key.
.Pp
.Nm
is part of the
.Xr monkeysphere 7
framework for providing a PKI for SSH.
.Sh CAVEATS
The keys produced by this process are stripped of all identifying
information, including certifications, self-signatures, etc.  This is
intentional, since ssh attaches no inherent significance to these
features.
.Pp
.Nm
will produce output for any requested RSA key.  This means, among
other things, that it will happily export revoked keys, unverifiable
keys, expired keys, etc.  Make sure you do your own key validation
before using this tool!
.Sh EXAMPLES
.Nm gpg \-\-export\-secret\-key $KEYID | openpgp2ssh $KEYID | ssh\-add \-c /dev/stdin
.Pp
This pushes the secret key into the active
.Xr ssh\-agent 1 . 
Tools such as 
.Xr ssh 1
which know how to talk to the 
.Xr ssh\-agent 1
can now rely on the key.
.Sh AUTHOR
.Nm
and this man page were written by Daniel Kahn Gillmor
<dkg@fifthhorseman.net>.
.Sh BUGS
.Nm
only works with RSA keys.  DSA keys are the only other key type
available in both OpenPGP and SSH, but they are currently unsupported
by this utility.
.Pp
.Nm
only accepts raw OpenPGP packets on standard input.  It does not
accept ASCII-armored input.
.Nm
Currently only exports into formats used by the OpenSSH.
It should support other key output formats, such as those used by
.Xr lsh 1
and
.Xr putty 1 .
.Pp
Secret key output is currently not passphrase-protected.
.Pp
.Nm
currently cannot handle passphrase-protected secret keys on input.
.Sh SEE ALSO
.Xr pem2openpgp 1 ,
.Xr monkeysphere 1 ,
.Xr monkeysphere 7 ,
.Xr ssh 1 ,
.Xr monkeysphere-authentication 8 ,
.Xr monkeysphere-host 8


Changelog
monkeysphere (0.36) unstable; urgency=low

  * keytrans no longer confuses user IDs across different keys (closes MS
    #2682)
  * fetch all available keys from keyserver instead of first 5 (closes MS
    #1046)
  * enable openpgp2pem for keytrans (Closes: #698383)
  * enable openpgp2spki as well

  [ Jonas Smedegaard ]
  * m-a gpg-cmd now takes its arguments as separate parameters, not as a
    single string.

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Thu, 11 Jul 2013 13:19:45 -0400

monkeysphere (0.35) upstream;

  * Remove reference to USE_VALIDATION_AGENT.
  * Fix ssh_proxycommand for marginal hosts (closes MS #2593)
  * GnuPG should always behave as --fixed-list-mode (closes MS #2587)

 -- Jameson Rollins <jrollins@finestructure.net>  Fri, 29 Oct 2010 20:21:54 -0400

monkeysphere (0.34) upstream;

  * fix keys-for-user so that it outputs proper authorized_keys lines
    (close MS #2550)
  * refactor key processing for key files, greatly reducing redundant code
    paths
  * update authorized_keys and known_hosts in temp filess that are
    atomically moved into place
  * don't fail if authorized_keys file not already present (Closes: 600644)
  * document CHECK_KEYSERVER in monkeysphere-authentication man page
    (close MS #2556)

 -- Jameson Rollins <jrollins@finestructure.net>  Tue, 26 Oct 2010 10:27:01 -0400

monkeysphere (0.33) upstream;

  [ Daniel Kahn Gillmor ]
  * defaulting MONKEYSPHERE_HASH_KNOWN_HOSTS to false
    (closes MS #2483)

  [ Jameson Rollins ]
  * fix security vulnerability is parsing userids in
    monkeysphere-authentication keys-for-user (Closes: #600304)
  * fix failure after first invalid key in monkeysphere-authentication
    keys-for-user (closes MS #2545)
  * ignore command options in monkeysphere-authentication keys-for-user

 -- Jameson Rollins <jrollins@finestructure.net>  Fri, 15 Oct 2010 18:05:18 -0400

monkeysphere (0.32) upstream;

  [ Jameson Rollins ]
  * Fix specification of install paths in all scripts and man pages
    (closes MS #2491)
  * Fix need for single argument to gpg_sphere (thanks Clint)
    (closes MS #442)
  * specify LC_ALL=C for all gpg calls
    (closes MS #2496)

  [ Micah Anderson ]
  * fix monkeysphere-host revoke-key, which never worked properly :(
  * add some debug output to monkeysphere-host publish-key
    (closes MS #2289)

  [ Clint Adams ]
  * add support for options to the authorized User IDs file.  Options that
    should apply to keys for a given User ID should be on
    whitespace-prefixed lines immediately following that User ID.
    (closes MS #440)

 -- Jameson Rollins <jrollins@finestructure.net>  Wed, 06 Oct 2010 17:41:09 -0400

monkeysphere (0.31) upstream;

  [ Daniel Kahn Gillmor ]
  * support x509 anchors for monkeysphere-host, allow shared anchor
    between m-h and m-a (closes MS #2288)
  * do not bail or fail on m-h publish-key if the admin interactively
    declines to publish one of the keys key (closes MS #1945)
  * report updated expiration date upon successful conclusion of m-h
    set-expire (closes MS #2291)
  * added some files in examples/ to demonstrate system integration
    with OpenSSH

  [ Jameson Rollins ]
  * add keys-for-user subcommand to monkeysphere-authentication

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Thu, 15 Jul 2010 19:20:35 -0400

monkeysphere (0.30) upstream;

  * changing tarball creation and packaging strategies
  * make non-ssh parts of monkeysphere work well when openssh is not
    installed; degrade ssh-specific parts gracefully when openssh is not
    installed.

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Sat, 17 Apr 2010 16:46:52 -0400

monkeysphere (0.29) upstream;

  * This is mainly a bugfix release
  * Fix man page typo about monkeysphere authorized_keys location
  * Monkeysphere should work properly even if the user has "armor" in
    their gpg.conf (closes MS #1625)
  * monkeysphere keys-for-userid now respects MONKEYSPHERE_CHECK_KEYSERVER
    environment variable (and defaults to true)
  * introduce monkeysphere sshfprs-for-userid (deprecates sshfpr), closes
    MS #1436
  * respect CHECK_KEYSERVER in more places (closes MS #1997)
  * warn on keyserver failures for monkeysphere-authentication (closes MS
    #1750)
  * avoid checking trustdb for monkeysphere-host (closes MS #1957)
  * allow monkeysphere-authentication to use hkps with trusted X.509 root
    certificate authorities in
    /etc/monkeysphere/monkeysphere-authentication-x509-anchors.crt

 -- Daniel Kahn Gillmor <dkg@fifthhorseman.net>  Sun, 14 Mar 2010 21:00:47 -0400

monkeysphere (0.28) upstream;

  * Major rework of monkeysphere-host to handle multiple host keys.  We
    also no longer assume ssh service keys.  monkeysphere-host is now a
    general-purpose host service OpenPGP key management UI.
  * Rename keys-from-userid command to more accurate keys-for-userid
  * separate upstream and debian changelogs

 -- Jameson Rollins <jrollins@finestructure.net>  Tue, 19 Jan 2010 13:50:31 -0500

monkeysphere (0.27) upstream;

  * fixed monkeysphere gen-subkey subcommand that was erroneously creating
    DSA subkeys due to unannounced change in gpg edit-key UI.  Now tests
    for gpg version (closes MS #1536)
  * add new monkeysphere keys-from-userid subcommand to output all
    acceptable keys for a given user ID literal

 -- Jameson Rollins <jrollins@finestructure.net>  Mon, 11 Jan 2010 20:54:21 -0500

monkeysphere (0.26) upstream;

  * add 'refresh-keys' subcommand to monkeysphere-authentication
  * improve marginal UI (closes MS #1141)
  * add MONKEYSPHERE_STRICT_MODES configuration to avoid
    permission-checking (closes MS #649)
  * test scripts use STRICT_MODES to avoid failure when built under /tmp
  * do permissions checks with a perl script instead of non-portable
    readlink GNUisms
  * bail on permissions check if we hit the home directory (helpful on Mac
    OS and other systems with loose /home or /Users (closes MS #675)


src/keytrans/openpgp2ssh.c
#include "gnutls-helpers.h"

#include <gnutls/openpgp.h>
#include <gnutls/x509.h>

/* for waitpid() */
#include <sys/types.h>
#include <sys/wait.h>

/* 
   Author: Daniel Kahn Gillmor <dkg@fifthhorseman.net>
   Date: 2008-06-12 13:47:41-0400
   License: GPL v3 or later

   monkeysphere key translator: execute this with an OpenPGP key on
   stdin, (please indicate the specific keyid that you want as the
   first argument if there are subkeys).  At the moment, only public
   keys and passphraseless secret keys work.

   For secret keys, it will spit out a PEM-encoded version of the key
   on stdout, which can be fed into ssh-add like this:

    gpg --export-secret-keys $KEYID | openpgp2ssh $KEYID | ssh-add -c /dev/stdin

   For public keys, it will spit out a single line of text that can
   (with some massaging) be used in an openssh known_hosts or
   authorized_keys file.  For example:

    echo server.example.org $(gpg --export $KEYID | openpgp2ssh $KEYID) >> ~/.ssh/known_hosts

   Requirements: I've only built this so far with GnuTLS v2.3.x.
   GnuTLS 2.2.x does not contain the appropriate functionality.

 */


/* FIXME: keyid should be const as well */
int convert_private_pgp_to_x509(gnutls_x509_privkey_t* output, const gnutls_openpgp_privkey_t* pgp_privkey, const unsigned char* keyfpr, unsigned int fprlen) {
  gnutls_datum_t m, e, d, p, q, u, g, y, x;
  gnutls_pk_algorithm_t pgp_algo;
  unsigned int pgp_bits;
  int ret;
  int subkeyidx;
  int subkeycount;
  int found = 0;
  unsigned char fingerprint[20];
  size_t fingerprint_length = sizeof(fingerprint); 

  init_datum(&m);
  init_datum(&e);
  init_datum(&d);
  init_datum(&p);
  init_datum(&q);
  init_datum(&u);
  init_datum(&g);
  init_datum(&y);
  init_datum(&x);

  subkeycount = gnutls_openpgp_privkey_get_subkey_count(*pgp_privkey);
  if (subkeycount < 0) {
    err(0,"Could not determine subkey count (got value %d)\n", subkeycount);
    return 1;
  }

  if ((keyfpr == NULL) && 
      (subkeycount > 0)) {
    err(0,"No key identifier passed in, but there were %d keys to choose from\n", subkeycount + 1);
    return 1;
  }

  if (keyfpr != NULL) {
    ret = gnutls_openpgp_privkey_get_fingerprint(*pgp_privkey, fingerprint, &fingerprint_length);
    if (ret) {
      err(0,"Could not get fingerprint (error: %d)\n", ret);
      return 1;
    }
    if (fprlen > fingerprint_length) {
      err(0, "Requested key identifier is longer than computed fingerprint\n");
      return 1;
    }
    if (fingerprint_length > fprlen) {
      err(0, "Only comparing last %d bits of key fingerprint\n", fprlen*8);
    }
  }
  if ((keyfpr == NULL) || (memcmp(fingerprint + (fingerprint_length - fprlen), keyfpr, fprlen) == 0)) {
    /* we want to export the primary key: */
    err(0,"exporting primary key\n");

    /* FIXME: this is almost identical to the block below for subkeys.
       This clumsiness seems inherent in the gnutls OpenPGP API,
       though.  ugh. */
    pgp_algo = gnutls_openpgp_privkey_get_pk_algorithm(*pgp_privkey, &pgp_bits);
    if (pgp_algo < 0) {
      err(0, "failed to get OpenPGP key algorithm (error: %d)\n", pgp_algo);
      return 1;
    }
    if (pgp_algo == GNUTLS_PK_RSA) {
      err(0,"OpenPGP RSA Key, with %d bits\n", pgp_bits);
      ret = gnutls_openpgp_privkey_export_rsa_raw(*pgp_privkey, &m, &e, &d, &p, &q, &u);
      if (GNUTLS_E_SUCCESS != ret) {
	err(0, "failed to export RSA key parameters (error: %d)\n", ret);
	return 1;
      }
      
    } else if (pgp_algo == GNUTLS_PK_DSA) {
      err(0,"OpenPGP DSA Key, with %d bits\n", pgp_bits);
      ret = gnutls_openpgp_privkey_export_dsa_raw(*pgp_privkey, &p, &q, &g, &y, &x);
      if (GNUTLS_E_SUCCESS != ret) {
	err(0,"failed to export DSA key parameters (error: %d)\n", ret);
	return 1;
      }
    }
    found = 1;
  } else {
    /* lets trawl through the subkeys until we find the one we want: */
    for (subkeyidx = 0; (subkeyidx < subkeycount) && !found; subkeyidx++) {
      ret = gnutls_openpgp_privkey_get_subkey_fingerprint(*pgp_privkey, subkeyidx, fingerprint, &fingerprint_length);
      if (ret) {
	err(0,"Could not get fingerprint of subkey with index %d (error: %d)\n", subkeyidx, ret);
	return 1;
      }
      if (fprlen > fingerprint_length) {
	err(0, "Requested key identifier is longer than computed fingerprint\n");
	return 1;
      }
      if (fingerprint_length > fprlen) {
	err(1, "Only comparing last %d bits of key fingerprint\n", fprlen*8);
      }
      if (memcmp(fingerprint + (fingerprint_length - fprlen), keyfpr, fprlen) == 0) {
	err(0,"exporting subkey index %d\n", subkeyidx);

	/* FIXME: this is almost identical to the block above for the
	   primary key. */
	pgp_algo = gnutls_openpgp_privkey_get_subkey_pk_algorithm(*pgp_privkey, subkeyidx, &pgp_bits);
	if (pgp_algo < 0) {
	  err(0,"failed to get the algorithm of the OpenPGP public key (error: %d)\n", pgp_algo);
	  return pgp_algo;
	} else if (pgp_algo == GNUTLS_PK_RSA) {
	  err(0,"OpenPGP RSA key, with %d bits\n", pgp_bits);
	  ret = gnutls_openpgp_privkey_export_subkey_rsa_raw(*pgp_privkey, subkeyidx, &m, &e, &d, &p, &q, &u);
	  if (GNUTLS_E_SUCCESS != ret) {
	    err(0,"failed to export RSA key parameters (error: %d)\n", ret);
	    return 1;
	  }
	} else if (pgp_algo == GNUTLS_PK_DSA) {
	  err(0,"OpenPGP DSA Key, with %d bits\n", pgp_bits);
	  ret = gnutls_openpgp_privkey_export_subkey_dsa_raw(*pgp_privkey, subkeyidx, &p, &q, &g, &y, &x);
	  if (GNUTLS_E_SUCCESS != ret) {
	    err(0,"failed to export DSA key parameters (error: %d)\n", ret);
	    return 1;
	  }
	}
	found = 1;
      }
    }
  }

  if (!found) {
    err(0,"Could not find key in input\n");
    return 1;
  }

  if (pgp_algo == GNUTLS_PK_RSA) {
    ret = gnutls_x509_privkey_import_rsa_raw (*output, &m, &e, &d, &p, &q, &u); 
    if (GNUTLS_E_SUCCESS != ret) {
      err(0, "failed to import RSA key parameters (error: %d)\n", ret);
      return 1;
    }
  } else if (pgp_algo == GNUTLS_PK_DSA) {
    ret = gnutls_x509_privkey_import_dsa_raw (*output, &p, &q, &g, &y, &x); 
    if (GNUTLS_E_SUCCESS != ret) {
      err(0,"failed to import DSA key parameters (error: %d)\n", ret);
      return 1;
    }
  } else {
    err(0,"OpenPGP Key was not RSA or DSA -- can't deal! (actual algorithm was: %d)\n", pgp_algo);
    return 1;
  }
  
  ret = gnutls_x509_privkey_fix(*output);
  if (ret != 0) {
    err(0,"failed to fix up the private key in X.509 format (error: %d)\n", ret);
    return 1; 
  }

  return 0;
}

/* FIXME: keyid should be const also */
int emit_public_openssh_from_pgp(const gnutls_openpgp_crt_t* pgp_crt, const unsigned char* keyfpr, size_t fprlen) {
  int ret;
  int subkeyidx;
  int subkeycount;
  int found = 0;
  gnutls_datum_t m, e, p, q, g, y, algolabel;
  unsigned int bits;
  gnutls_pk_algorithm_t algo;
  const gnutls_datum_t* all[5];
  const char* algoname;
  int mpicount;
  /* output_data must be at least 2 chars longer than the maximum possible
     algorithm name: */
  char output_data[20];

  unsigned char fingerprint[20];
  size_t fingerprint_length = sizeof(fingerprint); 

  /* variables for the output conversion: */
  int pipestatus;
  int pipefd, child_pid;
  char* const b64args[] = {"sh", "-c", "base64 | tr -c -d '[A-Za-z0-9=+/]'", NULL};

  init_datum(&m);
  init_datum(&e);
  init_datum(&p);
  init_datum(&q);
  init_datum(&g);
  init_datum(&algolabel);


  /* figure out if we've got the right thing: */
  subkeycount = gnutls_openpgp_crt_get_subkey_count(*pgp_crt);
  if (subkeycount < 0) {
    err(0,"Could not determine subkey count (got value %d)\n", subkeycount);
    return 1;
  }

  if ((keyfpr == NULL) && 
      (subkeycount > 0)) {
    err(0,"No key identifier passed in, but there were %d keys to choose from\n", subkeycount + 1);
    return 1;
  }

  if (keyfpr != NULL) {
    ret = gnutls_openpgp_crt_get_fingerprint(*pgp_crt, fingerprint, &fingerprint_length);
    if (ret) {
      err(0,"Could not get key fingerprint (error: %d)\n", ret);
      return 1;
    }
    if (fprlen > fingerprint_length) {
      err(0, "Requested key identifier is longer than computed fingerprint\n");
      return 1;
    }
    if (fingerprint_length > fprlen) {
      err(0, "Only comparing last %d bits of key fingerprint\n", fprlen*8);
    }
  }
  if ((keyfpr == NULL) || (memcmp(fingerprint + (fingerprint_length - fprlen), keyfpr, fprlen) == 0)) {
    /* we want to export the primary key: */
    err(0,"exporting primary key\n");

    /* FIXME: this is almost identical to the block below for subkeys.
       This clumsiness seems inherent in the gnutls OpenPGP API,
       though.  ugh. */
    algo = gnutls_openpgp_crt_get_pk_algorithm(*pgp_crt, &bits);
    if (algo < 0) {
      err(0,"failed to get the algorithm of the OpenPGP public key (error: %d)\n", algo);
      return algo;
    } else if (algo == GNUTLS_PK_RSA) {
      err(0,"OpenPGP RSA certificate, with %d bits\n", bits);
      ret = gnutls_openpgp_crt_get_pk_rsa_raw(*pgp_crt, &m, &e);
      if (GNUTLS_E_SUCCESS != ret) {
	err(0,"failed to export RSA certificate parameters (error: %d)\n", ret);
	return 1;
      }
    } else if (algo == GNUTLS_PK_DSA) {
      err(0,"OpenPGP DSA certificate, with %d bits\n", bits);
      ret = gnutls_openpgp_crt_get_pk_dsa_raw(*pgp_crt, &p, &q, &g, &y);
      if (GNUTLS_E_SUCCESS != ret) {
	err(0,"failed to export DSA certificate parameters (error: %d)\n", ret);
	return 1;
      }
    }
    found = 1;

  } else {
    /* lets trawl through the subkeys until we find the one we want: */
    for (subkeyidx = 0; (subkeyidx < subkeycount) && !found; subkeyidx++) {
      ret = gnutls_openpgp_crt_get_subkey_fingerprint(*pgp_crt, subkeyidx, fingerprint, &fingerprint_length);
      if (ret) {
	err(0,"Could not get fingerprint of subkey with index %d (error: %d)\n", subkeyidx, ret);
	return 1;
      }
      if (fprlen > fingerprint_length) {
	err(0, "Requested key identifier is longer than computed fingerprint\n");
	return 1;
      }
      if (fingerprint_length > fprlen) {
	err(1, "Only comparing last %d bits of key fingerprint\n", fprlen*8);
      }
      if (memcmp(fingerprint + (fingerprint_length - fprlen), keyfpr, fprlen) == 0) {
	err(0,"exporting subkey index %d\n", subkeyidx);

	/* FIXME: this is almost identical to the block above for the
	   primary key. */
	algo = gnutls_openpgp_crt_get_subkey_pk_algorithm(*pgp_crt, subkeyidx, &bits);
	if (algo < 0) {
	  err(0,"failed to get the algorithm of the OpenPGP public key (error: %d)\n", algo);
	  return algo;
	} else if (algo == GNUTLS_PK_RSA) {
	  err(0,"OpenPGP RSA certificate, with %d bits\n", bits);
	  ret = gnutls_openpgp_crt_get_subkey_pk_rsa_raw(*pgp_crt, subkeyidx, &m, &e);
	  if (GNUTLS_E_SUCCESS != ret) {
	    err(0,"failed to export RSA certificate parameters (error: %d)\n", ret);
	    return 1;
	  }
	} else if (algo == GNUTLS_PK_DSA) {
	  err(0,"OpenPGP DSA certificate, with %d bits\n", bits);
	  ret = gnutls_openpgp_crt_get_subkey_pk_dsa_raw(*pgp_crt, subkeyidx, &p, &q, &g, &y);
	  if (GNUTLS_E_SUCCESS != ret) {
	    err(0,"failed to export DSA certificate parameters (error: %d)\n", ret);
	    return 1;
	  }
	}
	found = 1;
	
      }
    }
  }

  if (!found) {
    err(0,"Could not find key in input\n");
    return 1;
  }

  /* if we made it this far, we've got MPIs, and we've got the
     algorithm, so we just need to emit the info */
  if (algo == GNUTLS_PK_RSA) {
    algoname = "ssh-rsa";
    mpicount = 3;

    all[0] = &algolabel;
    all[1] = &e;
    all[2] = &m;
  } else if (algo == GNUTLS_PK_DSA) {
    algoname = "ssh-dss";
    mpicount = 5;

    all[0] = &algolabel;
    all[1] = &p;
    all[2] = &q;
    all[3] = &g;
    all[4] = &y;
  } else {
    err(0,"Key algorithm was neither DSA nor RSA (it was %d).  Can't deal.  Sorry!\n", algo);
    return 1;
  }

  if (ret = datum_from_string(&algolabel, algoname), ret) {
    err(0,"couldn't label string (error: %d)\n", ret);
    return ret;
  }

  snprintf(output_data, sizeof(output_data), "%s ", algoname);

  pipefd = create_writing_pipe(&child_pid, b64args[0], b64args);
  if (pipefd < 0) {
    err(0,"failed to create a writing pipe (returned %d)\n", pipefd);
    return pipefd;
  }
    
  write(1, output_data, strlen(output_data));

  if (0 != write_data_fd_with_length(pipefd, all, mpicount)) {
    err(0,"was not able to write out RSA key data\n");
    return 1;
  }
  close(pipefd);
  if (child_pid != waitpid(child_pid, &pipestatus, 0)) {
    err(0,"could not wait for child process to return for some reason.\n");
    return 1;
  }
  if (pipestatus != 0) {
    err(0,"base64 pipe died with return code %d\n", pipestatus);
    return pipestatus;
  }

  write(1, "\n", 1);
  
  return 0;
}

int main(int argc, char* argv[]) {
  gnutls_datum_t data;
  int ret = 0;
  gnutls_x509_privkey_t x509_privkey;
  gnutls_openpgp_privkey_t pgp_privkey;
  gnutls_openpgp_crt_t pgp_crt;

  char output_data[10240];
  size_t ods = sizeof(output_data);
  
  unsigned char * fingerprint = NULL;
  size_t fpr_size;
  char * prettyfpr = NULL;

  init_gnutls();

  /* figure out what key we should be looking for: */
  if (argv[1] != NULL) {
    if (strlen(argv[1]) > 81) {
      /* safety check to avoid some sort of wacky overflow situation:
	 there's no reason that the key id should be longer than twice
	 a sane fingerprint (one byte between chars, and then another
	 two at the beginning and end) */
      err(0, "Key identifier is way too long.  Please use at most 40 hex digits.\n");
      return 1;
    }

    fpr_size = hexstring2bin(NULL, argv[1]);
    if (fpr_size > 40*4) {
      err(0, "Key identifier is longer than 40 hex digits\n");
      return 1;
    }
    /* since fpr_size is initially in bits: */
    if (fpr_size % 8 != 0) {
      err(0, "Please provide an even number of hex digits for the key identifier\n");
      return 1;
    }
    fpr_size /= 8;

    fingerprint = malloc(sizeof(unsigned char) * fpr_size);
    bzero(fingerprint, sizeof(unsigned char) * fpr_size);
    hexstring2bin(fingerprint, argv[1]);

    prettyfpr = malloc(sizeof(unsigned char)*fpr_size*2 + 1);
    if (prettyfpr != NULL) {
      hex_print_data(prettyfpr, fingerprint, fpr_size);
      prettyfpr[sizeof(unsigned char)*fpr_size*2] = '\0';
      err(1, "searching for key with fingerprint '%s'\n", prettyfpr);
      free(prettyfpr);
    }

    if (fpr_size < 4) {
      err(0, "You MUST provide at least 8 hex digits in any key identifier\n");
      return 1;
    }
    if (fpr_size < 8)
      err(0, "You should provide at least 16 hex digits in any key identifier (proceeding with %d digits anyway)\n", fpr_size*2);
   
  }

  
  init_datum(&data);

  /* slurp in the key from stdin */
  if (ret = set_datum_fd(&data, 0), ret) {
    err(0,"didn't read file descriptor 0\n");
    return 1;
  }


  if (ret = gnutls_openpgp_privkey_init(&pgp_privkey), ret) {
    err(0,"Failed to initialized OpenPGP private key (error: %d)\n", ret);
    return 1;
  }
  /* check whether it's a private key or a public key, by trying them: */
  if ((gnutls_openpgp_privkey_import(pgp_privkey, &data, GNUTLS_OPENPGP_FMT_RAW, NULL, 0) == 0) || 
      (gnutls_openpgp_privkey_import(pgp_privkey, &data, GNUTLS_OPENPGP_FMT_BASE64, NULL, 0) == 0)) {
    /* we're dealing with a private key */
    err(0,"Translating private key\n");
    if (ret = gnutls_x509_privkey_init(&x509_privkey), ret) {
      err(0,"Failed to initialize X.509 private key for output (error: %d)\n", ret);
      return 1;
    }
    
    ret = convert_private_pgp_to_x509(&x509_privkey, &pgp_privkey, fingerprint, fpr_size);

    gnutls_openpgp_privkey_deinit(pgp_privkey);
    if (ret)
      return ret;

    ret = gnutls_x509_privkey_export (x509_privkey,
				      GNUTLS_X509_FMT_PEM,
				      output_data,
				      &ods);
    if (ret == 0) {
      write(1, output_data, ods);
    }
    gnutls_x509_privkey_deinit(x509_privkey);
  
  } else {
    if (ret = gnutls_openpgp_crt_init(&pgp_crt), ret) {
      err(0,"Failed to initialized OpenPGP certificate (error: %d)\n", ret);
      return 1;
    }
    
    if ((gnutls_openpgp_crt_import(pgp_crt, &data, GNUTLS_OPENPGP_FMT_RAW) == 0) || 
	(gnutls_openpgp_crt_import(pgp_crt, &data, GNUTLS_OPENPGP_FMT_BASE64) == 0)) {
      /* we're dealing with a public key */
      err(0,"Translating public key\n");

      ret = emit_public_openssh_from_pgp(&pgp_crt, fingerprint, fpr_size);
      if (ret != 0) 
	return ret;
      
    } else {
      /* we have no idea what kind of key this is at all anyway! */
      err(0,"Input does not contain any form of OpenPGP key I recognize.\n");
      return 1;
    }
  }

  gnutls_global_deinit();
  free(fingerprint);
  return 0;
}
